### PR TITLE
Tiered Version Store

### PR DESCRIPTION
**Pull request by snazy**
TL;DR there's a [README](https://github.com/snazy/nessie/tree/tvs2-experimentation-space/versioned/tiered) that also contains a couple of ToDo's / Follow-Ups

Review notes: it's best to follow the individual commits - a bunch of those are rather "boring" (the massive-delete one, test-code adpotions).

The whole logic around commits, merges, transplants, fetching keys and values resides in
`AbstractDatabaseAdapter` and is shared across all kinds of database adapters.

Database adapters, for both transactional and non-transactional databases, have the database
specific implementations around the CAS-loop for non-transactional, catching
integrity-constraint-violations for transactional, the concrete physical data model, concrete read &
write implementations.

Logical Data model
==================

The `DatabaseAdapter` interface defines the functions needed by the tiered-version-store implementation to access the data.

Implementations of `DatabaseAdapter` are free to implement their own optimizations.

Non-transactional
-----------------

The logical data model shared by all database adapters consists of three entities:

* _Commit-log_ contains all commit-log-entries, identified by a deterministic hash. This is the same
  as for transactional databases.
* _Global-state-log_ contains all changes to Iceberg/Delta/Hive tables and views, identified by a
  rather random ID.
* _Global-pointer_ a single "table row" that points to the current _global-state-log_ and all HEADs
  for the named references. Consistent updates are guaranteed via a CAS operation on this entity
  comparing the HEAD of the global-state-log.

Transactional
-------------

The data for transactional is a classic relational data model that consists of three tables:

* _Commit-log_ contains all commit-log-entries, identified by a deterministic hash. This is the same
  as for non-transactional databases.
* _Named-references_ contains all named-references and their current HEAD, the latter is used to
  guarantee consistent updates.
* _Global-state_ contains the current global state for a contents-key. Consistent changes are
  guaranteed by tracking the hash of the contents of the global-state-value.

**************************************************************************************************************

Two major API changes
============================

1. Add global-state to Contents
2. Always use a named+reference + hash

Add global-state to Contents
-------------------------------------------------

To make potentially conflicting changes (DDL or DML) to a table safe, Nessie needs to validate
certain global table attributes and act as a kind of "source of truth" for these attributes.
Those attributes are:
* last-column-id
* current-schema-id
* last-assigned-partition-id
* current-snapshot-id

Nessie must ensure that changes to the same table in different Nessie branches use globally
unique values for the above attributes.

"Global contents" information is meant to be tracked by Nessie for all tables across all
branches/tags.

"(Per branch) contents" contains information that is relevant to the Nessie branch/tag.

The approach here is to introduce a new schema-type hierarchy `GlobalContents` next to
`Contents` plus a new `Operation.PutGlobal`. A commit-operation must then pass two operations
for the same `ContentsKey`.
* a `Operation.Put` with the `Contents` object and
* a `Operation.PutGlobal` with the expected `GlobalContents` object and the updated
  `GlobalContents` object.

It feels that a new exception next to `NessieConflictException` is not necessary and Nessie
can raise/return a `NessieConflictExcepiton` in case the change from "expected global state"
to "proposed global state" cannot succeed.

Although we have technically two "contents" objects (`IcebergTable` + `IcebergSnapshot`),
we do not have to return both objects from the `ContentsApi`, because the only global
attribute (table-manifest-location) is present in both "contents" objects.

Always use a named+reference + hash
-------------------------------------------------

For authz to work properly, to ensure that a provided hash is really accessible to the caller,
a bunch of changes had to be applied to the API calls.

This change is [Reviewable](https://reviewable.io/reviews/projectnessie/nessie/1638)

### CODE REVIEW

**Review by nastra — COMMENTED** [review]
did a first pass over the the first seven commits

File: versioned/tiered/nontx/src/main/java/org/projectnessie/versioned/tiered/nontx/GlobalStatePointer.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        .globalId(globalId)
+        .namedReferences(namedReferences)
+        .build();
+  }
+
+  default GlobalStatePointer update(Hash newGlobalHead, NamedRef ref, Hash toHead) {
+    List<Ref> curr = getNamedReferences();
+    List<Ref> newRefs = new ArrayList<>(curr.size() + 1);
+
+    if (toHead == null) {
```
[the remark this answers is not in the document]
Comment (snazy) [reply versioned/tiered/nontx/src/main/java/org/projectnessie/versioned/tiered/nontx/GlobalStatePointer.java:58]: yup, added a `delete()`

File: versioned/tiered/adapter/src/main/java/org/projectnessie/versioned/tiered/adapter/DatabaseAdapter.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+   * @param expectedHead expected HEAD in {@code branch}
+   * @param expectedStates map expected global-states by {@link Key}
+   * @param puts serialized put-operations
+   * @param unchanged unchanged content-keys
+   * @param deletes deleted content-keys
+   * @param operationsKeys all keys contained in {@code puts}, {@code unchanged}, {@code deletes}
+   * @param commitMetaSerialized serialized commit-metadata
+   * @return optimistically written commit-log-entry
+   */
+  Hash commit(
```
[the remark this answers is not in the document]
Comment (snazy) [reply versioned/tiered/adapter/src/main/java/org/projectnessie/versioned/tiered/adapter/DatabaseAdapter.java:77]: true, added a wrapper type

File: versioned/tiered/nontx/src/main/java/org/projectnessie/versioned/tiered/nontx/NonTxDatabaseAdapter.java [hunk 3]
```diff
@@ -0,0 +1,581 @@
...
+  }
+
+  @Override
+  public void assign(
+      NamedRef ref, Optional<Hash> expectedHash, NamedRef assignTo, Optional<Hash> assignToHash)
+      throws ReferenceNotFoundException, ReferenceConflictException {
+    try {
+      casOpLoop(
+          ref,
+          true,
```
[the remark this answers is not in the document]
Comment (snazy) [reply versioned/tiered/nontx/src/main/java/org/projectnessie/versioned/tiered/nontx/NonTxDatabaseAdapter.java:356]: Nice catch! Fixed and added some more docs around that.

**Comment by snazy** [comment]
Closing this one, all relevant changes merged to main
